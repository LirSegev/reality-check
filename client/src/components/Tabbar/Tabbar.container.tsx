import React from 'react';
import ReactDOM from 'react-dom';

import TabbarView from './Tabbar.view';

interface Props {
	tabs: NonEmptyArray<{
		tabTitle: string;
		content: JSX.Element;
		className?: string;
	}>;
	index?: number;
	onChange?: (e: { index: number }) => void;
}

interface State {
	index: number;
}

class TabbarContainer extends React.Component<Props, State> {
	constructor(props: Props) {
		super(props);

		this.state = {
			index: props.index || 0,
		};

		this._handleClick = this._handleClick.bind(this);
	}

	_barRef = React.createRef<HTMLDivElement>();
	componentDidMount() {
		this._barRef.current?.addEventListener('touchstart', e => {
			e.stopPropagation();
		});
	}

	componentDidUpdate(prevProps: Props, prevState: State) {
		const propsIndex = this.props.index;
		if (propsIndex !== undefined && propsIndex !== prevProps.index)
			this._changeTab(propsIndex);
	}

	_handleClick(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
		const index = Number(e.currentTarget.dataset.index);
		this._changeTab(index);
	}

	_changeTab(index: number) {
		this.setState({
			index,
		});

		// prettier-ignore
		const tabEl = (ReactDOM.findDOMNode(this) as Element)
			.querySelector( `[data-index="${index}"]` )!
		// @ts-ignore
		if (tabEl.scrollIntoViewIfNeeded) tabEl.scrollIntoViewIfNeeded(); // TODO: test with IOS

		if (this.props.onChange) this.props.onChange({ index });
	}

	render() {
		return (
			<TabbarView
				tabs={this.props.tabs}
				index={this.state.index}
				handleClick={this._handleClick}
				barRef={this._barRef}
			/>
		);
	}
}

export default TabbarContainer;
