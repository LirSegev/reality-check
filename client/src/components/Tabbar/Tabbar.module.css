.bar {
	width: 100%;
	height: calc(2.5em * 1.15);
	margin-top: calc((-2.5em * 0.15) / 2);
	overflow-x: scroll;
	display: flex;

	position: sticky;
	top: 0;
	z-index: 10;
	background-color: #efeff4;
}

.bar .button {
	min-width: fit-content;
	padding: 0 0.5em;
	height: 2.5em;
	line-height: 1em;
	flex-grow: 1;
	font-weight: bold;
	margin-top: calc((2.5em * 0.15) / 2);
}

.bar .button.active {
	z-index: 1;
	background-color: #cacaca;
	transform: scale(1.15);
}

/* 
	Button styling from onsen ui
*/
.bar button:focus {
	border-top: none;
	box-shadow: none;
	outline: 0;
}
.bar button {
	/* font-weight: 400; */
	box-sizing: border-box;
	background-clip: padding-box;
	padding: 0;
	margin: 0;
	/* font: inherit; */
	/* color: inherit; */
	background: transparent;
	border: none;
	/* line-height: normal; */
	cursor: default;
	-webkit-user-select: none;
	-moz-user-select: none;
	user-select: none;
	text-overflow: ellipsis;
	white-space: nowrap;
	overflow: hidden;
	position: relative;
	display: inline-block;
	text-decoration: none;
	padding: 0;
	/* height: 49px; */
	letter-spacing: 0;
	/* color: #999; */
	vertical-align: top;
	background-color: transparent;
	border-top: none;
	/* width: 100%; */
	/* font-weight: 400; */
	/* line-height: 49px; */
}
