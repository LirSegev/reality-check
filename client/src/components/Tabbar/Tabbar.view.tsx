import React from 'react';

import styles from './Tabbar.module.css';

export interface Props {
	tabs: NonEmptyArray<{
		tabTitle: string;
		content: JSX.Element;
		className?: string;
	}>;
	index: number;
	handleClick: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

const TabbarView: React.FC<Props> = props => {
	const tabs = props.tabs.map((tab, index) => {
		const classes = [styles.button];
		if (tab.className) classes.push(tab.className);

		if (index === props.index) classes.push(styles.active);

		return (
			<button
				className={classes.join(' ')}
				onClick={props.handleClick}
				data-index={index}
				key={`tabButton_${index}-${tab.tabTitle}`}
			>
				{tab.tabTitle}
			</button>
		);
	});

	return (
		<React.Fragment>
			<div className={styles.bar}>{tabs}</div>
			<div className={styles.content} data-testid="content">
				{props.tabs[props.index].content}
			</div>
		</React.Fragment>
	);
};

export default TabbarView;
