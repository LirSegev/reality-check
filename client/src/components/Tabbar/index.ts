import Tabbar from './Tabbar.view';

export default Tabbar;
