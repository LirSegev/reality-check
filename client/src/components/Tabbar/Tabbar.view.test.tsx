import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

import styles from './Tabbar.module.css';
import Tabbar from './Tabbar.view';

afterEach(cleanup);

it('renders', () => {
	const { asFragment } = render(
		<Tabbar
			tabs={[
				{
					tabTitle: 'lir',
					content: <div></div>,
				},
			]}
			handleClick={() => {}}
			index={0}
		/>
	);

	expect(asFragment()).toMatchSnapshot();
});

it('displays tabTitle', () => {
	const { getByText } = render(
		<Tabbar
			tabs={[
				{
					tabTitle: 'lir segev',
					content: <div></div>,
				},
				{
					tabTitle: 'tom segev',
					content: <div></div>,
				},
			]}
			handleClick={() => {}}
			index={0}
		/>
	);

	expect(getByText('lir segev')).toBeVisible();
	expect(getByText('tom segev')).toBeVisible();
});

it('highlights active tab', () => {
	const { getByText } = render(
		<Tabbar
			tabs={[
				{
					tabTitle: 'lir segev',
					content: <div></div>,
				},
				{
					tabTitle: 'tom segev',
					content: <div></div>,
				},
			]}
			handleClick={() => {}}
			index={1}
		/>
	);

	expect(getByText('tom segev')).toHaveClass(styles.active);
});

it('calls handleClick() correctly', () => {
	const handleClick = jest.fn(
		(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
			const index = e.currentTarget.dataset.index;
			expect(index).toBe('0');
		}
	);

	const { getByText } = render(
		<Tabbar
			tabs={[
				{
					tabTitle: 'lir segev',
					content: <div></div>,
				},
			]}
			handleClick={handleClick}
			index={0}
		/>
	);

	fireEvent.click(getByText('lir segev'));
	expect(handleClick).toBeCalledTimes(1);
});

it('displays content', () => {
	const { getByTestId } = render(
		<Tabbar
			tabs={[
				{
					tabTitle: 'lir segev',
					content: <div>Lir segev's content</div>,
				},
			]}
			handleClick={() => {}}
			index={0}
		/>
	);
	expect(getByTestId('content')).toContainHTML(
		"<div>Lir segev's content</div>"
	);
});
