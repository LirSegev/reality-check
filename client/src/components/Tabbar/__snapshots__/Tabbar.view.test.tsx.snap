// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="bar"
  >
    <button
      class="button active"
      data-index="0"
    >
      lir
    </button>
  </div>
  <div
    class="content"
    data-testid="content"
  >
    <div />
  </div>
</DocumentFragment>
`;
